### ISSUE TITLE
Windows - chromeSpawnHelper doesn't work with Electron

### ISSUE THREAD

**Issue by greg9504** [microsoft/vscode-chrome-debug#407]
- Debugger for Chrome Version:
3.0.0
- VS Code Version:
Version 1.11.1
Commit d9484d12b38879b7f4cdd1150efeb2fd2c1fbf39
Date 2017-04-06T13:58:12.954Z
Shell 1.4.6
Renderer 53.0.2785.143
Node 6.5.0
- Electron Version
1.5.1
- Webpack Version
2.3.3
Steps to reproduce:

1.  I updated the Debugger for Chrome to version 3.0.0.  My Electron application that was previously debuggable no longer works in either the Renderer or Main process.

I'm using the following launch configuration:
```
{
            "name": "Debug Dev Renderer Process",
            "type": "chrome",
            "request": "launch",
            "runtimeExecutable": "${workspaceRoot}/node_modules/.bin/electron.cmd",           
            "runtimeArgs": [
              "${workspaceRoot}/dist",
              "--enable-logging",
              "--remote-debugging-port=9222"
            ],
            "sourceMaps": true,   
            "trace": "verbose",       
            "webRoot": "${workspaceRoot}"
                
        }
```
again this worked.  Now when I attempt to start a debugging session I see a command window briefly appear and then noting.  My Electron application does not start and I get the following error inside VS Code:

> Cannot connect to runtime process, timeout after 10000 ms - (reason: Cannot connect to the target: connect ECONNREFUSED 127.0.0.1:9222).

Using .script and http://localhost:9222/json both produce nothing because the application is not running.

If I start the application from the command prompt and pass it the debugging flags like:
```
D:\>D:\devtools\js\tutorials\angular-electron-dream-starter-master\node
_modules\.bin\electron.cmd "D:\devtools\js\tutorials\angular-electron-dream-starter-master\dist" --enable-logging --remote-debugging-port=9222
```
http://localhost:9222/json  produces:
[ {
   "description": "",
   "devtoolsFrontendUrl": "/devtools/inspector.html?ws=localhost:9222/devtools/page/7a092244-2fd1-4aea-b69a-f9e300d682fa",
   "id": "7a092244-2fd1-4aea-b69a-f9e300d682fa",
   "title": "Angular Electron Dream Starter by Colin Skow &amp; @AngularClass",
   "type": "page",
   "url": "file:///D:/devtools/js/tutorials/angular-electron-dream-starter-master/dist/index.html#/",
   "webSocketDebuggerUrl": "ws://localhost:9222/devtools/page/7a092244-2fd1-4aea-b69a-f9e300d682fa"
}, {
   "description": "",
   "devtoolsFrontendUrl": "/devtools/inspector.html?ws=localhost:9222/devtools/page/fd8120cf-9147-41e8-a087-e7f4f3e6097d",
   "id": "fd8120cf-9147-41e8-a087-e7f4f3e6097d",
   "title": "chrome-extension://redux-devtools/_generated_background_page.html",
   "type": "page",
   "url": "chrome-extension://redux-devtools/_generated_background_page.html",
   "webSocketDebuggerUrl": "ws://localhost:9222/devtools/page/fd8120cf-9147-41e8-a087-e7f4f3e6097d"
} ]

So if my app would start it appears it should be debuggable.  The problem is why isn't it starting from the VS Code chrome launcher ?

Here is the trace output:

> OS: win32 ia32
> Adapter node: v6.5.0 ia32
> vscode-chrome-debug-core: 3.14.17
> 5:19:05 PM, 4/12/2017
> debugger-for-chrome: 2.7.3
> From client: initialize({"clientID":"vscode","adapterID":"chrome","pathFormat":"path","linesStartAt1":true,"columnsStartAt1":true,"supportsVariableType":true,"supportsVariablePaging":true,"supportsRunInTerminalRequest":true})
> To client: {"seq":0,"type":"response","request_seq":1,"command":"initialize","success":true,"body":{"exceptionBreakpointFilters":[{"label":"All Exceptions","filter":"all","default":false},{"label":"Uncaught Exceptions","filter":"uncaught","default":true}],"supportsConfigurationDoneRequest":true,"supportsSetVariable":true,"supportsConditionalBreakpoints":true,"supportsCompletionsRequest":true,"supportsHitConditionalBreakpoints":true,"supportsRestartFrame":true,"supportsExceptionInfoRequest":true,"supportsRestartRequest":true}}
> From client: launch({"name":"Debug Dev Renderer Process","type":"chrome","request":"launch","runtimeExecutable":"d:\\devtools\\js\\tutorials\\angular-electron-dream-starter-master/node_modules/.bin/electron.cmd","runtimeArgs":["d:\\devtools\\js\\tutorials\\angular-electron-dream-starter-master/dist","--enable-logging","--remote-debugging-port=9222"],"sourceMaps":true,"trace":"verbose","webRoot":"d:\\devtools\\js\\tutorials\\angular-electron-dream-starter-master"})
> To client: {"seq":0,"type":"event","event":"output","body":{"category":"telemetry","output":"debugStarted","data":{"request":"launch","args":["name","type","request","runtimeExecutable","runtimeArgs","sourceMaps","trace","webRoot","sourceMapPathOverrides","skipFileRegExps"]}}}
> Discovering targets via http://127.0.0.1:9222/json
> [chromeSpawnHelper] spawn('d:\devtools\js\tutorials\angular-electron-dream-starter-master/node_modules/.bin/electron.cmd', ["--remote-debugging-port=9222","--no-first-run","--no-default-browser-check","d:\\devtools\\js\\tutorials\\angular-electron-dream-starter-master/dist","--enable-logging","--remote-debugging-port=9222"])
> 
> got chrome PID: 2860
> HTTP GET failed: Error: connect ECONNREFUSED 127.0.0.1:9222
> Discovering targets via http://127.0.0.1:9222/json
> HTTP GET failed: Error: connect ECONNREFUSED 127.0.0.1:9222
> Discovering targets via http://127.0.0.1:9222/json
> HTTP GET failed: Error: connect ECONNREFUSED 127.0.0.1:9222
> Discovering targets via http://127.0.0.1:9222/json
> HTTP GET failed: Error: connect ECONNREFUSED 127.0.0.1:9222
> Discovering targets via http://127.0.0.1:9222/json
> HTTP GET failed: Error: connect ECONNREFUSED 127.0.0.1:9222
> Discovering targets via http://127.0.0.1:9222/json
> HTTP GET failed: Error: connect ECONNREFUSED 127.0.0.1:9222
> Discovering targets via http://127.0.0.1:9222/json
> HTTP GET failed: Error: connect ECONNREFUSED 127.0.0.1:9222
> Discovering targets via http://127.0.0.1:9222/json
> HTTP GET failed: Error: connect ECONNREFUSED 127.0.0.1:9222
> Discovering targets via http://127.0.0.1:9222/json
> HTTP GET failed: Error: connect ECONNREFUSED 127.0.0.1:9222
> To client: {"seq":0,"type":"response","request_seq":2,"command":"launch","success":false,"message":"Cannot connect to runtime process, timeout after 10000 ms - (reason: Cannot connect to the target: connect ECONNREFUSED 127.0.0.1:9222).","body":{"error":{"id":2010,"format":"Cannot connect to runtime process, timeout after {_timeout} ms - (reason: {_error}).","variables":{"_error":"Cannot connect to the target: connect ECONNREFUSED 127.0.0.1:9222","_timeout":"10000"}}}}
> From client: disconnect({"restart":false})
> 

Note it reports the version as:
debugger-for-chrome: 2.7.3
yet I have 3.0.0 installed and the file is written to: 
C:\Users\userx\.vscode\extensions\msjsdiag.debugger-for-chrome-3.0.0

Is there any way to go back to version 2.7.3 of the extension?
Thanks

**Comment by auchenberg** [comment]
Looks like the `--remote-debugging-port=9222` param is being added multiple times in `
spawn('C:/Users/greg9_000/AppData/Roaming/npm/electron.cmd', ["--remote-debugging-port=9222","--no-first-run","--no-default-browser-check","d:\devtools\js\tutorials\angular-electron-dream-starter-master/dist","--enable-logging","--remote-debugging-port=9222"])
`

Can you try to remove it from `runtimeArgs`? Electron is today an unsupported scenario.

**Comment by greg9504** [comment]
Hello,
Thanks for the reply.  That unfortunately didn't work

> [chromeSpawnHelper] spawn('d:\devtools\js\tutorials\angular-electron-dream-starter-master/node_modules/.bin/electron.cmd', ["--remote-debugging-port=9222","--no-first-run","--no-default-browser-check","d:\\devtools\\js\\tutorials\\angular-electron-dream-starter-master/dist","--enable-logging"])


I even tried remove all runtimeArgs.   

Is there any way to go back to the old version of the extension?

**Comment by greg9504** [comment]
OK, I managed to downgrade to 2.7.3 and debugging is back working with my Electron application.
For anyone who runs into this you can dl the 2.7.3 version from:
https://microsoft.gallery.vsassets.io/_apis/public/gallery/publisher/microsoft/extension/msjsdiag.debugger-for-chrome/2.7.3/assetbyname/Microsoft.VisualStudio.Services.VSIXPackage

then rename it to so it ends in .vsix

Uninstall the 3.0.0 version of the extension.  Reload.  Install 2.7.3 version using the  "Install from VSIX..." command in the Extensions View command drop-down.

See https://code.visualstudio.com/docs/editor/extension-gallery  under Common Questions for further details.

**Comment by auchenberg** [comment]
@greg9504 Can I get you to poste the trace output when using 2.7.3? This would easy the debugging to find out what broke.

**Comment by greg9504** [comment]
Yes of course, that was stupid of me not to include it... 
Important parts below, full file attached.

> OS: win32 ia32
> Adapter node: v6.5.0 ia32
> vscode-chrome-debug-core: 3.14.14
> 5:59:34 PM, 4/12/2017
> debugger-for-chrome: 2.7.3
> From client: 
initialize({"clientID":"vscode","adapterID":"chrome","pathFormat":"path","linesStartAt1":true,"columnsStartAt1":true,"supportsVariableType":true,"supportsVariablePaging":true,"supportsRunInTerminalRequest":true})
> To client: {"seq":0,"type":"response","request_seq":1,"command":"initialize","success":true,"body":{"exceptionBreakpointFilters":[{"label":"All Exceptions","filter":"all","default":false},{"label":"Uncaught Exceptions","filter":"uncaught","default":true}],"supportsConfigurationDoneRequest":true,"supportsSetVariable":true,"supportsConditionalBreakpoints":true,"supportsCompletionsRequest":true,"supportsHitConditionalBreakpoints":true,"supportsRestartFrame":true,"supportsExceptionInfoRequest":true,"supportsRestartRequest":true}}
> From client: launch({"name":"Debug Dev Renderer Process","type":"chrome","request":"launch","runtimeExecutable":"d:\\devtools\\js\\tutorials\\angular-electron-dream-starter-master/node_modules/.bin/electron.cmd","runtimeArgs":["d:\\devtools\\js\\tutorials\\angular-electron-dream-starter-master/dist","--enable-logging","--remote-debugging-port=9222"],"sourceMaps":true,"trace":"verbose","webRoot":"d:\\devtools\\js\\tutorials\\angular-electron-dream-starter-master"})
> To client: {"seq":0,"type":"event","event":"output","body":{"category":"telemetry","output":"debugStarted","data":{"request":"launch","args":["name","type","request","runtimeExecutable","runtimeArgs","sourceMaps","trace","webRoot","sourceMapPathOverrides","skipFileRegExps"]}}}
> spawn('d:\devtools\js\tutorials\angular-electron-dream-starter-master/node_modules/.bin/electron.cmd', ["--remote-debugging-port=9222","--no-first-run","--no-default-browser-check","d:\\devtools\\js\\tutorials\\angular-electron-dream-starter-master/dist","--enable-logging","--remote-debugging-port=9222"])
> Discovering targets via http://127.0.0.1:9222/json
> To client: {"seq":0,"type":"event","event":"output","body":{"category":"telemetry","output":"targetCount","data":{"numTargets":2}}}
> Warning: Found more than one valid target page. Attaching to the first one. Available pages: ["file:///D:/devtools/js/tutorials/angular-electron-dream-starter-master/dist/index.html","chrome-extension://redux-devtools/_generated_background_page.html"]
> Attaching to target: {"description":"","devtoolsFrontendUrl":"/devtools/inspector.html?ws=127.0.0.1:9222/devtools/page/6f5145a5-3591-4866-8002-f93a3f65be70","id":"6f5145a5-3591-4866-8002-f93a3f65be70","title":"index.html","type":"page","url":"file:///D:/devtools/js/tutorials/angular-electron-dream-starter-master/dist/index.html","webSocketDebuggerUrl":"ws://127.0.0.1:9222/devtools/page/6f5145a5-3591-4866-8002-f93a3f65be70"}
> WebSocket Url: ws://127.0.0.1:9222/devtools/page/6f5145a5-3591-4866-8002-f93a3f65be70
> → To target: "{\"id\":1,\"method\":\"Debugger.setBlackboxPatterns\",\"params\":{\"patterns\":[\"^chrome-extension:.*\"]}}"
> → To target: "{\"id\":2,\"method\":\"Console.enable\"}"
> → To target: "{\"id\":3,\"method\":\"Debugger.enable\"}"
> → To target: "{\"id\":4,\"method\":\"Runtime.enable\"}"
> → To target: "{\"id\":5,\"method\":\"Runtime.runIfWaitingForDebugger\"}"
> → To target: "{\"id\":6,\"method\":\"Runtime.run\"}"
> → To target: "{\"id\":7,\"method\":\"Page.enable\"}"
> → To target: "{\"id\":8,\"method\":\"Network.enable\",\"params\":{}}"
> ← From target: {"id":1,"result":{}}
> ← From target: {"method":"Console.messageAdded","params":{"message":{"source":"console-api","level":"log","text":"loading react devtools","url":"file:///D:/devtools/js/tutorials/angular-electron-dream-starter-master/dist/main.bundle.js","line":435,"column":13}}}
> ← From target: {"method":"Console.messageAdded","params":{"message":{"source":"console-api","level":"log","text":"Initial App State","url":"file:///D:/devtools/js/tutorials/angular-electron-dream-starter-master/dist/main.bundle.js","line":340,"column":21}}}
> ← From target: {"method":"Console.messageAdded","params":{"message":{"source":"console-api","level":"log","text":"Angular is running in the development mode. Call enableProdMode() to enable the production mode.","url":"file:///D:/devtools/js/tutorials/angular-electron-dream-starter-master/dist/vendor.dll.js","line":603,"column":13}}}
> ← From target: {"method":"Console.messageAdded","params":{"message":{"source":"console-api","level":"log","text":"hello `Home` component","url":"file:///D:/devtools/js/tutorials/angular-electron-dream-starter-master/dist/main.bundle.js","line":597,"column":17}}}
> ← From target: {"id":2,"result":{}}
> ← From target: 


[vscode-chrome-debug.zip](https://github.com/Microsoft/vscode-chrome-debug/files/917995/vscode-chrome-debug.zip)

**Comment by auchenberg** [comment]
The difference seems to be:

2.7.3:
```javascript
launch({
            "name": "Debug Dev Renderer Process",
            "type": "chrome",
            "request": "launch",
            "runtimeExecutable": "d:\devtools\js\tutorials\angular-electron-dream-starter-master/node_modules/.bin/electron.cmd",
            "runtimeArgs": [
"d:\devtools\js\tutorials\angular-electron-dream-starter-master/dist",
 "--enable-logging", 
"--remote-debugging-port=9222"]
```

3.0.0:
```javascript
spawn('d:\devtools\js\tutorials\angular-electron-dream-starter-master/node_modules/.bin/electron.cmd', [
  "--remote-debugging-port=9222", 
  "--no-first-run", 
  "--no-default-browser-check", 
  "d:\devtools\js\tutorials\angular-electron-dream-starter-master/dist", 
  "--enable-logging", 
  "--remote-debugging-port=9222"
])
```

**Comment by greg9504** [comment]
I don't think that is it.  You quoted the "launch" log from 2.7.3 and the "spawn" log form 3.0.0.  Looking both "spawn" logs, they are identical.  It would appear to be something that changed that isn't obvious in the logs.  

I do see 
[chromeSpawnHelper]
In the 3.0.0 log that is not in the 2.7.3 log.  I've attached the 3.0.0 log so anyone who wants to diff can, easy to see the differences.

[vscode-chrome-debug_3.0.0.txt.zip](https://github.com/Microsoft/vscode-chrome-debug/files/918037/vscode-chrome-debug_3.0.0.txt.zip)

**Comment by roblourens** [comment]
Thanks for the logs, I can repro this. The problem is with the chromeSpawnHelper which I added to keep Chrome from dying when the debug adapter shuts down. I have no idea why Electron has a problem with this on Windows. Investigating...

**Comment by roblourens** [comment]
I still have no idea why this is happening. I don't know whether to disable it when runtimeExecutable is set (which sucks for people using it to launch Chrome Canary or something), or add yet another option to disable it. Maybe disable for runtimeExecutable in the short term.

**Comment by roblourens** [comment]
It's like it's running my electron main.js as a normal node app. Giving up for now - I just published 3.0.1 with a workaround for this.

**Comment by roblourens** [comment]
https://github.com/Microsoft/vscode-chrome-debug/commit/81db25f1d925985febc62af03328001a611b9c07
